### PR TITLE
Add a new plugin to handle HP M2xfd monitors

### PR DESCRIPTION
**Pull request by gportay**
Added support for genesys logic devices.

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/genesys/fu-genesys-scaler-device.c [hunk 1]
```diff
@@ -0,0 +1,1806 @@
...
+						   NULL,
+						   error)) {
+			g_prefix_error(error,
+				       "error reading flash at 0x%06x: ",
+				       fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		fu_progress_set_percentage_full(progress,
+						i * self->transfer_size + fu_chunk_get_data_sz(chk),
+						bufsz);
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-scaler-device.c:923]: can you use `fu_progress_set_steps(progress, chunks->len);` and then just do `fu_progress_step_done(progress);` here?

File: plugins/genesys/fu-genesys-scaler-device.c [hunk 2]
```diff
@@ -0,0 +1,1806 @@
...
+									 fu_chunk_get_address(chk),
+									 error)) {
+			g_prefix_error(error,
+				       "error erasing flash at address 0x%06x: ",
+				       fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		fu_progress_set_percentage_full(progress,
+						i * self->sector_size + fu_chunk_get_data_sz(chk),
+						bufsz);
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-scaler-device.c:1229]: same with `fu_progress_step_done(progress);`

File: plugins/genesys/fu-genesys-scaler-device.c [hunk 3]
```diff
@@ -0,0 +1,1806 @@
...
+			    buf,
+			    bufsz,
+			    0, /* src */
+			    bufsz,
+			    error))
+		return FALSE;
+
+	chunks =
+	    fu_chunk_array_mutable_new(data, datasz, addr + sizeof(data1), 0, self->transfer_size);
+	for (guint i = 0; i < chunks->len; i++) {
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-scaler-device.c:1292]: no progress updates?

File: plugins/genesys/fu-genesys-scaler-device.c [hunk 4]
```diff
@@ -0,0 +1,1806 @@
...
+	guint addr = 0x000000;
+	g_autofree guint8 *buf = NULL;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_DUAL_IMAGE))
+		addr = 0x200000;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-scaler-device.c:1497]: can you try this and get the daemon-provided values? then you can remove `GUESSED`

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 5]
```diff
@@ -0,0 +1,1333 @@
...
+		fu_common_dump_raw(G_LOG_DOMAIN,
+				   "Static info",
+				   (guint8 *)&self->static_ts,
+				   sizeof(FuGenesysStaticToolString));
+	}
+	if (self->static_ts.tool_string_version != 0xff) {
+		gchar rev[3] = {0};
+		guint64 tmp;
+
+		/* [TODO] move this to the quirk file */
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:746]: do you still want to do this? Maybe we just remove the TODO for now as we have hardcoded constants

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 6]
```diff
@@ -0,0 +1,1333 @@
...
+				     guint bufsz,
+				     GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	FuGenesysWaitFlashRegisterHelper helper = {.reg = 5, .expected_val = 0};
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	chunks =
+	    fu_chunk_array_new(buf, bufsz, start_addr, self->flash_block_size, self->flash_rw_size);
+	for (guint i = 0; i < chunks->len; i++) {
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:1019]: FuProgress here?

File: plugins/genesys/fu-genesys-usbhub-firmware.c [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+fu_genesys_usbhub_firmware_verify(const guint8 *buf, gsize bufsz, guint16 code_size, GError **error)
+{
+	guint16 fw_checksum, checksum;
+
+	/* get checksum */
+	if (!fu_common_read_uint16_safe(buf, bufsz, 0x5FFE, &fw_checksum, G_BIG_ENDIAN, error))
+		return FALSE;
+
+	/* calculate checksum */
+	checksum = fu_common_sum16(buf, code_size - sizeof(checksum));
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-firmware.c:31]: is code_size always > sizeof(checksum)?

**Review by superm1 — COMMENTED** [review]
> plugins/genesys/tests/genesys-usbhub.bin

Is this redistributable?  If not we should be putting it in [fwupd-test-firmware](https://github.com/fwupd/fwupd-test-firmware)

**Comment by hughsie** [comment]
> Is this redistributable

I assumed it was built using the .builder.xml file...

**Comment by gportay** [comment]
Thanks for your review, I will address the point this week.

**Comment by gportay** [comment]
> > > Is this redistributable
> > 
> > 
> > I assumed it was built using the .builder.xml file...
> 
> Well want to make sure :)

Well I do not know where it come from :sweat: I will investigate...

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 8]
```diff
@@ -170,6 +175,8 @@ fu_genesys_usbhub_device_read_flash(FuGenesysUsbhubDevice *self,
 				       fu_chunk_get_address(chk));
 			return FALSE;
 		}
+		if (progress)
```
[the remark this answers is not in the document]
Comment (gportay) [reply plugins/genesys/fu-genesys-usbhub-device.c:175]: Oops, adding a fixup right now

[the remark this answers is not in the document]
Comment (gportay) [reply plugins/genesys/fu-genesys-usbhub-device.c]: done

**Comment by gportay** [comment]
@hughsie do you prefer I rebase and push-force the changes or keeping this for the very end?

**Comment by gportay** [comment]
Looks like the fedora .spec file has to be updated afterward.

```
 RPM build errors:
    Installed (but unpackaged) file(s) found:
   /usr/lib/debug/usr/lib64/fwupd-plugins-5/libfu_plugin_genesys.so-1.7.5-0.1alpha.fc35.x86_64.debug
   /usr/lib64/fwupd-plugins-5/libfu_plugin_genesys.so
```

I wonder how far it is relevant to check for the creation of the package :/

**Comment by gportay** [comment]
And the LGTM analysis fails (at least) because of the symbol `g_usb_version_string()` was added couple of commits after [0.3.0](https://github.com/hughsie/libgusb/commits/fc5d3abb961aff5fc8882c70569f251c544b322f) and the container uses libgusb-0.3.0.

The plugin needs [commit](https://github.com/hughsie/libgusb/commit/4e118c154dde70e196c4381bd97790a9413c3552) at runtime upstreamed in libgusb-3.8.0.

```
[2022-02-08 18:47:42] [build-stdout] /usr/bin/ld: plugins/genesys/libfu_plugin_genesys.so.p/fu-genesys-usbhub-device.c.o: in function `fu_genesys_usbhub_device_setup':
[2022-02-08 18:47:42] [build-stdout] fu-genesys-usbhub-device.c:(.text+0x18e4): undefined reference to `g_usb_device_get_spec'
[2022-02-08 18:47:42] [build-stdout] /usr/bin/ld: fu-genesys-usbhub-device.c:(.text+0x1907): undefined reference to `g_usb_version_string'
[2022-02-08 18:47:42] [build-stdout] collect2: error: ld returned 1 exit status
```

I wonder if the image can be upgraded.

Also, I guess we can remove the preprocessor check for libgusb version or just add a warning at runtime.

**Comment by hughsie** [comment]
> Looks like the fedora .spec file has to be updated afterward

Fixed.

**Comment by hughsie** [comment]
> because of the symbol g_usb_version_string() was added

Lets just not use it then :) Fixed.

**Comment by gportay** [comment]
Oh, doomed.

The symbol [g_usb_device_get_spec](https://github.com/hughsie/libgusb/commits/13233c38eed954baf0ca3a95a23bd28ae02c8ba7) is also missing from 0.3.0. It needs 0.3.1 at least.

```
[2022-02-08 20:47:05] [build-stdout] /usr/bin/ld: plugins/genesys/libfu_plugin_genesys.so.p/fu-genesys-usbhub-device.c.o: in function `fu_genesys_usbhub_device_setup':
[2022-02-08 20:47:05] [build-stdout] fu-genesys-usbhub-device.c:(.text+0x18e4): undefined reference to `g_usb_device_get_spec'
[2022-02-08 20:47:05] [build-stdout] collect2: error: ld returned 1 exit status
```

File: plugins/genesys/fu-genesys-scaler-device.c (hunk 4 above)
Comment (hughsie) [reply plugins/genesys/fu-genesys-scaler-device.c:1508]: Does the daemon not give you the values required, e.g. `raw timing data was { 0.002, 21.413, 0.595 } -- steps were set as [ 10 80 10 ] but should have been [ 1 97 2 ] at ../plugins/foo/fu-foo-device.c:3172`

**Comment by hughsie** [comment]
> do you prefer I rebase and push-force the changes

When you're ready can you squash all these down into one commit, and then i think it LGTM.

File: plugins/genesys/fu-genesys-scaler-device.c (hunk 4 above)
Comment (gportay) [reply plugins/genesys/fu-genesys-scaler-device.c:1508]: I am giving a try. does it need the --verbose flag?

**Comment by gportay** [comment]
> > do you prefer I rebase and push-force the changes
> 
> When you're ready can you squash all these down into one commit, and then i think it LGTM.

Thanks for your commits.

I will squash the commits and push force and rebase the commits on top of main.

I will give you the green light then.

File: plugins/genesys/fu-genesys-scaler-device.c (hunk 4 above)
Comment (hughsie) [reply plugins/genesys/fu-genesys-scaler-device.c:1508]: It might if you're running with -Dsupported=true

**Comment by gportay** [comment]
@hughsie Do you mind if I add the `X-VideoDisplay` category? that would mach with the `video-display` icon. in fwupd and lvfs-website

**Comment by hughsie** [comment]
> Do you mind if I add the X-VideoDisplay category

Makes sense to me; can you do it in a new PR please as that should be easier to review! :)

**Review by hughsie — APPROVED** [review]

**Pull request merged**